roll:
  localizationMode: false
  saveKeyframeMap: true 
  savePose: true # for path recording purposes /reloc gt, save it to saveMapDirectory
  savePCD: true
  saveMapDirectory:  /home/lzq/ROLL_ws/logs/roll2/map_pcd           
  saveKeyframeMapDirectory:  /home/lzq/ROLL_ws/logs/roll2/keyframes # for relocalization
  saveRawCloud: false
  debugMode: true
  # initialGuess: [-0.017984725855664, 0.012800135555456, -0.239421644654498, 75.3285, 106.723,-3.231]
  saveLog: true
  mapLoaded: false
  relocSuccess: false
  useOdom: false  

  alti0: 270.0
  lati0: 42.293227
  longi0: -83.709657
  # Visualization
  globalMapVisualizationSearchRadius: 1000.0    # meters, global map visualization radius(index would overflow if too big)
  globalMapVisualizationPoseDensity: 10.0       # 10 by default, meters, global map visualization keyframe density
  globalMapVisualizationLeafSize: 1.0         # 1 by default, meters, global map visualization cloud density

    # Surrounding map
  surroundingkeyframeAddingDistThreshold: 1.0   # 1.0 by default,meters, regulate keyframe adding threshold
  surroundingkeyframeAddingAngleThreshold: 0.2  # 0.2 by default,radians, regulate keyframe adding threshold
  surroundingKeyframeDensity: 1.0               # meters, downsample surrounding keyframe poses, why differs from add threshold???
  surroundingKeyframeSearchRadius: 50.0         # 20 is not okay; meters, within n meters scan-to-map optimization (when loop closure disabled)
  # Topics
  pointCloudTopic: cloud_registered_body              # Point cloud data
  imuTopic: imu_6                        # IMU data
  # odomTopic: "odometry/imu"                   # IMU pre-preintegration odometry, same frequency as IMU
  odomTopic: odom
  gpsTopic: fix1                # GPS odometry topic from navsat, see module_navsat.launch file
  gtTopic: ground_truth1 # use either gps or fix, not both!!!

  # GPS Settings
  gpsCovThreshold: 2.0                        # m^2, threshold for using GPS data
  poseCovThreshold: 1.0                      # 0.1 would be too frequent, 1.0 by default, m^2, threshold for using GPS data
  alti0: 270.0
  lati0: 42.293227
  longi0: -83.709657
  
  # Frames
  lidarFrame: "lidar_link"
  mapFrame: "odom"
  # NCLT HDL-32: 
  sensor: velodyne                            # lidar sensor type, either 'velodyne' or 'ouster'
  N_SCAN: 128                                 ## number of lidar channel (i.e., 16, 32, 64, 128)
  Horizon_SCAN: 1800                          ## lidar horizontal resolution (Velodyne:1800, Ouster:512,1024,2048,robosense: 2000)
  downsampleRate: 0.5                           #  default: 1. Downsample your data if too many points. i.e., 16 = 64 / 4, 16 = 16 / 1

  lidarMinRange: 0.2                          # default: 1.0, minimum lidar range to be used
  lidarMaxRange: 150.0                      ## default: 1000.0, maximum lidar range to be used

  # LOAM feature threshold
  edgeThreshold: 0.1 # changed the way getting smoothness
  surfThreshold: 0.1
  edgeFeatureMinValidNum: 10
  surfFeatureMinValidNum: 100

  # voxel filter paprams
  odometrySurfLeafSize: 0.4                     # default: 0.4 - outdoor, 0.2 - indoor
  mappingCornerLeafSize: 0.2                    # default: 0.2 - outdoor, 0.1 - indoor
  mappingSurfLeafSize: 0.4                      # default: 0.4 - outdoor, 0.2 - indoor

  # CPU Params
  numberOfCores: 4                              # number of cores for mapping optimization

  # Loop closure
  loopClosureEnableFlag: true
  loopClosureFrequency: 1.0                    # Hz, regulate loop closure constraint add frequency
  surroundingKeyframeSize: 50                   # submap size (when loop closure enabled)
  historyKeyframeSearchRadius: 50.0             # meters, key frame that is within n meters from current pose will be considerd for loop closure
  historyKeyframeSearchTimeDiff: 30.0           # 30 by default,seconds, key frame that is n seconds older will be considered for loop closure
  historyKeyframeSearchNum: 25                  # number of hostory key frames will be fused into a submap for loop closure
  # 0.3 by default, but for indoor it can go wrong, actually mostly for good matches, fitnessScore is around 0.03 ;
  #icp threshold, the smaller the better alignment
  historyKeyframeFitnessScore: 0.2              


